common/utils/include/claragenomics/utils/allocator.hpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * NVIDIA CORPORATION and its licensors retain all intellectual property
+ * and proprietary rights in and to this software, related documentation
+ * and any modifications thereto.  Any use, reproduction, disclosure or
+ * distribution of this software and related documentation without an express
+ * license agreement from NVIDIA CORPORATION is strictly prohibited.
+ */
+
+#pragma once
+#include <cub/util_allocator.cuh>
+
+namespace claragenomics
+{
+
+/**
+ * @brief Interface for a asynchronous device allocator.
+ */
+class DeviceAllocator
+{
+public:
+    virtual void* allocate(std::size_t n, cudaStream_t stream) = 0;
+
+    virtual void deallocate(void* p, std::size_t, cudaStream_t stream) = 0;
+};
+
+/**
+ * @brief Interface for a asynchronous host allocator.
+ */
+class HostAllocator
+{
+public:
+    virtual void* allocate(std::size_t n, cudaStream_t stream) = 0;
+
+    virtual void deallocate(void* p, std::size_t, cudaStream_t strean) = 0;
+};
+
+/**
+ * @brief Default cudaMalloc/cudaFree based device allocator 
+ */
+class CudaMallocAllocator : public DeviceAllocator
+{
+public:
+    void* allocate(std::size_t n, cudaStream_t) override
+    {
+        void* ptr = 0;
+        CGA_CU_CHECK_ERR(cudaMalloc(&ptr, n));
+        return ptr;
+    }
+    void deallocate(void* p, std::size_t, cudaStream_t) override
+    {
+        cudaError_t status = cudaFree(p);
+        if (cudaSuccess != status)
+        {
+            // deallocate should not throw execeptions which is why CGA_CU_CHECK_ERR is not used.
+        }
+    }
+};
+
+/**
+ * @brief A simple caching allocator for device memory allocations.
+ */
+class CachingDeviceAllocator : public DeviceAllocator
+{
+public:
+    CachingDeviceAllocator(size_t max_cached_bytes = 1e9)
```

Comment:
What happens if this is not the case? A non-cached memory gets allocated?

`Anchor` arrays are usually larger than 1GB.

Not saying this value should be increased, we can check the performance impact, but would like to make sure it does not just fail to allocate.

Reply:
Yes, a non-cached memory gets allocated. Larger allocations that cannot satisfy the allocator criteria are allocated normally and are simply freed when they are deallocated instead of being cached.